### PR TITLE
feat: remove Js.String2, use labeled arguments for Js.String

### PR DESCRIPTION
**Pull request by anmonteiro**
part of the unification proposed in https://github.com/melange-re/melange/issues/731

this needs the [melange v3 long-lived branch](https://github.com/reasonml/reason-react/pull/821) in reason-react to compile the template projects.

### CODE REVIEW

File: jscomp/test/dist/jscomp/test/js_string_test.js [hunk 1]
```diff
@@ -492,22 +493,27 @@ var suites_1 = {
...
+                                                                                      var arg = /(#)(:)?/;
                                                                                       return {
                                                                                               TAG: /* Eq */0,
                                                                                               _0: [
                                                                                                 "a",
                                                                                                 "#",
                                                                                                 undefined
                                                                                               ],
-                                                                                              _1: "a#b#:c".split(/(#)(:)?/, 3)
+                                                                                              _1: (function (param) {
```
Comment (anmonteiro) [inline jscomp/test/dist/jscomp/test/js_string_test.js:514]: expected since the test uses `|>`.

**Comment by anmonteiro** [comment]
Before we merge this, we should add a test for string interpolation that fails in the current PR state. There's still a mention of `String2` in the ppx:

https://github.com/melange-re/melange/blob/ac03e246cdd921ade9b64c1f27f410dc59315921/ppx/utf8_string.ml#L463-L464

File: jscomp/runtime/js_string.ml [hunk 2]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
+(** [codePointAt s ~pos] returns the code point at position [pos] within string
+    [s] as a [Some] value. The return value handles code points greater than or
+    equal to [0x10000]. If there is no code point at the given position, the
+    function returns [None].
 
 {[
-  codePointAt 1 {js|¿😺?|js} = Some 0x1f63a
-  codePointAt 5 "abc" = None
+  codePointAt {js|¿😺?|js} 1 = Some 0x1f63a
+  codePointAt "abc" 5 = None
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:143]: I think `pos` is missing here.

File: jscomp/runtime/js_string.ml [hunk 3]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
-  charCodeAt 0 {js|😺|js} returns 0xd83d
-  codePointAt 0 {js|😺|js} returns Some 0x1f63a
+  charCodeAt {js|😺|js} ~pos:0 returns 0xd83d
+  codePointAt {js|😺|js} ~pos:0 returns Some 0x1f63a
 ]}
-
 *)
 
-(** [codePointAt n s] returns the code point at position [n] within string [s] as a [Some] value. The return value handles code points greater than or equal to [0x10000]. If there is no code point at the given position, the function returns [None].
+external codePointAt : t -> pos:int -> int option = "codePointAt"
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:134]: Maybe can rename `pos` to `index`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt#index

File: jscomp/runtime/js_string.ml [hunk 4]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
-  endsWith "world!" "Hello, World!" = false;; (* case-sensitive *)
-  endsWith "World" "Hello, World!" = false;; (* exact match *)
+  endsWith "Hello, World!" ~suffix:"World!" = true;;
+  endsWith "Hello, World!" ~suffix:"world!" = false;; (* case-sensitive *)
+  endsWith "Hello, World!" ~suffix:"World" = false;; (* exact match *)
 ]}
 *)
 
-(** [endsWithFrom ending len str] returns [true] if the first [len] characters of [str] end with [ending], [false] otherwise. If [n] is greater than or equal to the length of [str], then it works like [endsWith]. (Honestly, this should have been named [endsWithAt], but oh well.)
+external endsWithFrom : t -> suffix:t -> len:int -> bool = "endsWith"
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:181]: Like in the Array PR, these functions can be unified, so there's a single `endsWith` with optional labelled args that match [MDN spec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith):

```ocaml
external endsWith : u -> ?searchString:u -> ?endPosition:int -> unit -> bool
  = "endsWith"
[@@mel.send]
```

File: jscomp/runtime/js_string.ml [hunk 5]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
-  indexOf "xyz" "bookseller" = -1;;
+  indexOfFrom "bookseller" ~sub:"ok" ~pos:1 = 2;;
+  indexOfFrom "bookseller" ~sub:"sell" ~pos:2 = 4;;
+  indexOfFrom "bookseller" ~sub:"sell" ~pos:5 = -1;;
 ]}
 *)
 
-external lastIndexOf : t -> int = "lastIndexOf"
-[@@mel.send.pipe: t]
+external lastIndexOf : t -> sub:t -> int = "lastIndexOf"
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:252]: s/`sub`/`searchString`?

File: jscomp/runtime/js_string.ml [hunk 6]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
-  lastIndexOf "xyz" "abcdefg" = -1;;
+  lastIndexOf "bookseller" ~sub:"ok" = 2;;
+  lastIndexOf "beekeeper" ~sub:"ee" = 4;;
+  lastIndexOf "abcdefg" ~sub:"xyz" = -1;;
 ]}
 *)
 
-external lastIndexOfFrom : t -> int -> int = "lastIndexOf"
-[@@mel.send.pipe: t]
+external lastIndexOfFrom : t -> sub:t -> pos:int -> int = "lastIndexOf"
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:267]: Same comment about simplifying.

File: jscomp/runtime/js_string.ml [hunk 7]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
 *)
-external normalize : t = "normalize"
-[@@mel.send.pipe: t]
-(** ES2015 *)
 
-external normalizeByForm : t -> t = "normalize"
-[@@mel.send.pipe: t]
-(**
-  [normalize str form] (ES2015) returns the normalized Unicode string using the specified form of normalization, which may be one of:
+external normalizeByForm : t -> form:[ `NFC | `NFD | `NFKC | `NFKD ] -> t
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:341]: Nice!

File: jscomp/runtime/js_string.ml [hunk 8]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
   }
 
-  @see <https://www.unicode.org/reports/tr15/tr15-45.html> Unicode technical report for details
+  @see <https://www.unicode.org/reports/tr15/tr15-45.html> Unicode technical
+  report for details
 *)
 
-(**
-  [repeat n s] returns a string that consists of [n] repetitions of [s]. Raises [RangeError] if [n] is negative.
+external repeat : t -> n:int -> t = "repeat"
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:358]: `count`? :grimacing:

File: jscomp/runtime/js_string.ml [hunk 9]
```diff
@@ -106,326 +101,363 @@ external get : t -> int -> t = ""
...
+[@@mel.send]
+(** [repeat s n] returns a string that consists of [n] repetitions of [s].
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:360]: Missing labelled arg in `repeat s n`

File: jscomp/runtime/js_string.ml [hunk 10]
```diff
@@ -434,221 +466,229 @@ let () = Js.log replaced (* prints "42" *)
...
+  splitByRe "has:no:match" ~regexp:[%re "/[,;]/"] = [|"has:no:match"|];;
 ]};
 *)
 
-external splitByReAtMost : Js_re.t -> limit:int -> t option array = "split"
-[@@mel.send.pipe: t]
-(**
-  [splitByReAtMost regex ~limit: n str] splits the given [str] at every occurrence of [regex] and returns an
-  array of the first [n] resulting substrings. If [n] is negative or greater than the number of substrings, the array will contain all the substrings.
+external splitByReAtMost : t -> regexp:Js_re.t -> limit:int -> t option array
```
Comment (jchavarri) [inline jscomp/runtime/js_string.ml:571]: Can fold with `splitByRe` by using optional labelled arg for `limit`.

**Review by jchavarri — COMMENTED** [review]
Nice! I didn't go through all cases where a simplification can be applied, as I want to know your thoughts about that suggestion first.

**Comment by anmonteiro** [comment]
> I didn't go through all cases where a simplification can be applied, as I want to know your thoughts about that suggestion first.

I like your suggestion. Check my reply in https://github.com/melange-re/melange/pull/895#discussion_r1394549955

File: jscomp/runtime/js_string.ml (hunk 3 above)
Comment (anmonteiro) [reply jscomp/runtime/js_string.ml:134]: I used `~pos` because that's what OCaml's `StdLabels` uses.

File: jscomp/runtime/js_string.ml (hunk 6 above)
Comment (anmonteiro) [reply jscomp/runtime/js_string.ml]: so turns out that `lastIndexOf` is one that we can't simplify like the others! the undefined argument influences the result.

File: jscomp/runtime/js_string.ml (hunk 5 above)
Comment (anmonteiro) [reply jscomp/runtime/js_string.ml]: StdLabels also uses `sub`, so I figured it'd be easier for OCaml folks.

Comment (jchavarri) [reply jscomp/runtime/js_string.ml]: This is a strange framework to me, considering they are bindings to JS APIs, not a Stdlib.

Comment (anmonteiro) [reply jscomp/runtime/js_string.ml]: OK. I don't have a strong opinion. Let me know what you think of https://github.com/melange-re/melange/pull/908

**Pull request merged**